### DISCUSSION TITLE
[Discussion] Tricky use case - public user profiles

### DISCUSSION THREAD

**Discussion by alexweininger** [appwrite/appwrite#1252] [General]
## Problem

It doesn't seem like there is a good way to get a list of users client side. 

There is no way to do it with client side account API. At first I thought I could just use the server-side API with only `users:read` permissions. 

However there is a major issue with this. All the users data is exposed this way. Things like emails, and preferences I don't necessarily want to be exposed.

And so if I want to store public user profiles, I would need to store them in the database? I would probably setup functions to listen to the account events to keep it updated.

## Question

Is the solution outlined above for public user profiles acceptable and good enough, or should we explore or push for features that make this problem easier for the developer?

## My thoughts

Some feature ideas I came up with for making this easier for the developer:

1. Adding an account property like `publicData` that is accessible from the client?
2. Maybe by default it's either empty or just has their `userId`.
3. The users can also control if this property is public or private.

I'm interested to hear @appwrite/core opinion as well as other Appwrite users/devs.

**Comment by eldadfux** [comment]
The reason we don't allow access to the users list from the client-side is related to endless privacy and security concerns that might popup with such implementation.

I think implementing a `publicData` attribute or a dedicated endpoint/s is a bit too coupled with a specific use-case and might be a lower abstraction layer from where Appwrite is. I also believe this use-case can be easily achieved with existing Appwrite APIs. We have to remember that even the number of users an app has can be a very sensitive information, and there is no way to allow someone to iterate over that list without exposing this information.

I see 3 main options here to allow this use-case:
1. Create a function that listens to user registration and creates a profile document for each user on a dedicated collection in the Appwrite Database
2. Create a document for the user profile directly from the client-side, once the user completes registration and creates the first session.
3. Create the profile page in a custom server side, or create a proxy that exposes the things you want from the Users API.

For me both the 1st and 2nd options makes most sense, and seems like the `Appwrite way` to do that.

**Reply by alexweininger** [reply]
Great! 1. and 2. are how I'm accomplishing this already. And I agree with all your points.

**Reply by alexweininger** [reply]
@eldadfux 
To help with this use case, would it be possible to have a document rule type of `userId`. Which is a text type except it validates that the text is a valid existing userId?

**Reply by eldadfux** [reply]
This is something we are actively investigating as part of the db upgrades we are working on. We are trying to implement a new validator for multiple Appwrite resources (user, file, collection, documnet, and more), might look something like: `users/[USER_ID]`

**Reply by alexweininger** [reply]
Yeah something like that, or being able to override the documentId, and set it to something like a userId would be useful.
